File: lure_link_flutter/lib/domains/entity/carpool.dart
import 'dart:core';
import 'package:lure_link_flutter/domains/value_object/carpool_status.dart';

class Carpool {
  int id;
  int organizerId;
  String organizerNickName;
  DateTime startTime;
  DateTime endTime;
  DateTime aplDeadline;
  String departurePoint;
  String departurePrefecture;
  String departureMunicipality;
  String destinationPoint;
  String destinationPrefecture;
  String destinationMunicipality;
  int budget;
  int maxParticipant;
  int currentParticipant;
  CarPoolStatus status;
  String? description;

  Carpool(
      this.id,
      this.organizerId,
      this.organizerNickName,
      this.startTime,
      this.endTime,
      this.aplDeadline,
      this.departurePoint,
      this.departurePrefecture,
      this.departureMunicipality,
      this.destinationPoint,
      this.destinationPrefecture,
      this.destinationMunicipality,
      this.budget,
      this.maxParticipant,
      this.currentParticipant,
      this.status,
      this.description,
  );

  Carpool.fromJson(Map<String, dynamic> json)
      : this(
          json['id'] as int,
          json['organizerId'] as int,
          json['organizerNickName'] as String,
          json['startTime'] as DateTime,
          json['endTime'] as DateTime,
          json['aplDeadline'] as DateTime,
          json['departurePoint'] as String,
          json['departurePrefecture'] as String,
          json['departureMunicipality'] as String,
          json['destinationPoint'] as String,
          json['destinationPrefecture'] as String,
          json['destinationMunicipality'] as String,
          json['budget'] as int,
          json['maxParticipant'] as int,
          json['currentParticipant'] as int,
          CarPoolStatus.parse(json['status']),
          json['description'] as String?,
        );

  Map<String, Object?> toJson() {
    return {
      'id':id,
      'organizerId':organizerId,
      'organizerNickName':organizerNickName,
      'startTime':startTime,
      'endTime':endTime,
      'aplDeadline':aplDeadline,
      'departurePoint':departurePoint,
      'departurePrefecture':departurePrefecture,
      'departureMunicipality':departureMunicipality,
      'destinationPoint':destinationPoint,
      'destinationPrefecture':destinationPrefecture,
      'destinationMunicipality':destinationMunicipality,
      'budget':budget,
      'maxParticipant':maxParticipant,
      'currentParticipant':currentParticipant,
      'status':status.display,
    };
  }
}

File: api/src/entity/car_pool.rs
use crate::domain::domain_object::budget::Budget;
use crate::domain::domain_object::carpool_status::CarPoolStatus;
use crate::domain::domain_object::fishing_point::PointName;
use crate::domain::domain_object::id::Id;
use crate::domain::domain_object::ja_timestamp::JaTimeStamp;
use crate::domain::domain_object::municipality::Municipality;
use crate::domain::domain_object::prefecture::Prefecture;
use crate::entity::user::User;
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub struct CarPool {
    pub id: Id,
    pub organizer: User,
    pub start_time: JaTimeStamp,
    pub end_time: JaTimeStamp,
    pub apl_deadline: JaTimeStamp,
    pub departure: Point,
    pub destination: Point,
    pub budget: Budget,
    pub max_participant: i16,
    pub current_participant: i16,
    pub status: CarPoolStatus,
    pub description: String,
}

#[cfg(test)]
impl Default for CarPool {
    fn default() -> Self {
        Self {
            id: 1i64.try_into().unwrap(),
            organizer: Default::default(),
            start_time: JaTimeStamp::default(),
            end_time: JaTimeStamp::default(),
            apl_deadline: JaTimeStamp::default(),
            departure: Default::default(),
            destination: Default::default(),
            budget: 1000i32.try_into().unwrap(),
            max_participant: 3,
            current_participant: 0,
            status: CarPoolStatus::Applying,
            description: "".to_string(),
        }
    }
}

#[derive(Debug, Serialize, Deserialize, Clone, PartialEq)]
pub struct Point {
    pub prefecture: Prefecture,
    pub municipality: Municipality,
    pub point_name: PointName,
}

impl Default for Point {
    fn default() -> Self {
        Self {
            prefecture: Prefecture::Hokkaido,
            municipality: "municipality".to_string().try_into().unwrap(),
            point_name: "a".to_string().try_into().unwrap(),
        }
    }
}

#[derive(Serialize, Deserialize, Clone, Debug, PartialEq)]
pub struct CreateCarPool {
    pub start_time: JaTimeStamp,
    pub end_time: JaTimeStamp,
    pub apl_deadline: JaTimeStamp,
    pub departure_prefecture_id: Id,
    pub departure_municipality_id: Id,
    pub departure_point: PointName,
    pub destination_prefecture_id: Id,
    pub destination_municipality_id: Id,
    pub destination_point: PointName,
    pub budget: Budget,
    pub max_participant: i16,
    pub description: String,
}

#[cfg(test)]
impl Default for CreateCarPool {
    fn default() -> Self {
        Self {
            start_time: JaTimeStamp::default(),
            end_time: JaTimeStamp::default(),
            apl_deadline: JaTimeStamp::default(),
            departure_prefecture_id: 1i64.try_into().unwrap(),
            departure_municipality_id: 1i64.try_into().unwrap(),
            departure_point: "a".to_string().try_into().unwrap(),
            destination_prefecture_id: 1i64.try_into().unwrap(),
            destination_municipality_id: 1i64.try_into().unwrap(),
            destination_point: "a".to_string().try_into().unwrap(),
            budget: 1000i32.try_into().unwrap(),
            max_participant: 0,
            description: "".to_string(),
        }
    }
}
